File: StudyplusAndroidSDK/src/main/java/jp/studyplus/android/sdk/service/auth/AuthTransit.java
package jp.studyplus.android.sdk.service.auth;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;

/**
 * Opens an Activity for a SDK client to be authorized.
 */
public class AuthTransit {
	private final Activity activity;
	private final int requestCode;

	public static AuthTransit from(Activity activity){
		return new AuthTransit(activity, 0);
	}

	private AuthTransit(Activity activity, int requestCode) {
		this.activity = activity;
		this.requestCode = requestCode;
	}

	public AuthTransit setRequestCode(int requestCode){
		return new AuthTransit(activity, requestCode);
	}

	public void startActivity(String consumerKey, String consumerSecret){
		Intent intent = new Intent();
		intent.setAction(Intent.ACTION_VIEW);
		intent.putExtra("consumer_key", consumerKey);
		intent.putExtra("consumer_secret", consumerSecret);
		intent.setData(Uri.parse("studyplus://auth/external/start"));
		activity.startActivityForResult(intent, requestCode);
	}
}

File: studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/internal/auth/AuthTransit.kt
package jp.studyplus.android.sdk.internal.auth

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import androidx.core.net.toUri
import jp.studyplus.android.sdk.R

internal class AuthTransit
/**
 * @param consumerKey for API
 * @param consumerSecret for API
 */
constructor(
    private val consumerKey: String,
    private val consumerSecret: String
) {
    companion object {
        private const val EXTRA_CONSUMER_KEY = "consumer_key"
        private const val EXTRA_CONSUMER_SECRET = "consumer_secret"

        private val URI_STORE = "market://details?id=jp.studyplus.android.app".toUri()
        private val INTENT_STORE = Intent(Intent.ACTION_VIEW, URI_STORE)
    }

    fun start(activity: Activity, requestCode: Int) {
        if (isInstalledStudyplus(activity).not()) {
            openGooglePlayStore(activity)
            return
        }

        val intent = Intent()
        intent.action = Intent.ACTION_VIEW
        intent.putExtra(EXTRA_CONSUMER_KEY, consumerKey)
        intent.putExtra(EXTRA_CONSUMER_SECRET, consumerSecret)
        intent.data = Uri.parse(activity.getString(R.string.app_custom_scheme))
        activity.startActivityForResult(intent, requestCode)
    }

    private fun openGooglePlayStore(context: Context) {
        context.startActivity(INTENT_STORE)
    }

    private fun isInstalledStudyplus(context: Context): Boolean {
        val packages = context.packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
        return packages.any { it.packageName == "jp.studyplus.android.app" }
    }
}
